R/CallbackSetLRScheduler.R

```diff
@@ -0,0 +1,208 @@
+#'   * "batch" - Step after each batch
+#'
+#'
+#' @export
+CallbackSetLRScheduler = R6Class("CallbackSetLRScheduler",
+  inherit = CallbackSet,
+  public = list(
+    scheduler_fn = NULL,
+    scheduler = NULL,
+    #' @description
+    #' Creates a new instance of this [R6][R6::R6Class] class.
+    initialize = function(.scheduler, ...) {
+      self$scheduler_fn = .scheduler
+      private$.scheduler_args = list(...)
+    },
+    #' @description
+    #' Creates the scheduler using the optimizer from the context
+    on_begin = function() {
+      self$scheduler = invoke(self$scheduler_fn, optimizer = self$ctx$optimizer, .args = private$.scheduler_args)
+    },
+    #' @description
+    #' Depending on the scheduler, step after each epoch
+    on_epoch_end = function() {
+      # TODO: ensure that this happens after optimizer$step()
+      # https://blogs.rstudio.com/ai/posts/2020-10-19-torch-image-classification/#training
+      # but for now let's hope that it does
+      self$scheduler$step()
+    }
+    # TODO: add batches
+  ),
+  private = list(
+    .scheduler_args = NULL
+  )
+)
+
+# TODO: determine whether we should set ranges and such even when torch does not
+
+# some of the schedulers accept lists
+# so they can treat different parameter groups differently
+check_class_or_list = function(x, classname) {
+  if (test_class(x, classname)) {
+    return(TRUE)
+  } else if (is.list(x)) {
+    if (some(!map_lgl(x, test_class(x, classname)))) {
+      return(paste0("One of the elements is not of class ", classname))
+    }
+  } else {
+    return(paste0("x must be of class ", classname, " or a list of objects of class ", classname))
+  }
+}
+
+# begin built-in torch LR schedulers
+
+#' @include TorchCallback.R
+mlr3torch_callbacks$add("lr_scheduler_cosine_annealing", function() {
+  TorchCallback$new(
+    callback_generator = CallbackSetLRScheduler,
+    param_set = ps(
+      T_max = p_int(tags = c("train", "required")),
+      eta_min = p_dbl(default = 0, lower = 0, tags = "train"),
+      last_epoch = p_int(default = -1, tags = "train"),
+      verbose = p_lgl(default = FALSE, tags = "train")
+    ),
+    id = "lr_scheduler",
+    label = "Learning Rate Scheduler",
+    man = "mlr3torch::mlr_callback_set.lr_scheduler",
+    additional_args = torch::lr_cosine_annealing
+  )
+})
+
+#' @include TorchCallback.R
+mlr3torch_callbacks$add("lr_scheduler_lambda", function() {
+  TorchCallback$new(
+    callback_generator = CallbackSetLRScheduler,
+    param_set = ps(
+      lr_lambda = p_uty(tags = c("train"), custom_check = function(x) check_class_or_list(x, "function")), # TODO: assert fn or list of fns
+      last_epoch = p_int(default = -1, lower = -1, tags = "train"),
+      verbose = p_lgl(default = FALSE, tags = "train")
+    ),
+    id = "lr_scheduler",
+    label = "Learning Rate Scheduler",
+    man = "mlr3torch::mlr_callback_set.lr_scheduler",
+    additional_args = lr_lambda
+  )
+})
```

Comment:
For the custom callback, we infer the parameters as the formalArgs` of the scheduling function.


```r
as_lr_scheduler = function(x) {
  # x is a lr scheduler function
  TorchCallback$new(
    callback_generator = CallbackSerLRScheduler,
    param_set = inferps(x),
    ...,
    additional_args = list(.scheduler = x)
    )
    ```
    
    If we had a scheduler that has arguments `a, b`, `inferps` creates a parameter set with two parameters `a` and `b` that are both of type `ParamUty`.